src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -203,7 +202,39 @@ irr_glsl_BSDFAnisotropicParams irr_glsl_calcBSDFAnisotropicParams(in irr_glsl_BS
 #endif
 )";
     }
+    static std::string getDiffuseFresnelCorrectionFactor(const std::string&)
+    {
+        return
+R"(#ifndef _IRR_BSDF_BRDF_DIFFUSE_FRESNEL_CORRECTION_INCLUDED_
+#define _IRR_BSDF_BRDF_DIFFUSE_FRESNEL_CORRECTION_INCLUDED_
+
+vec3 irr_glsl_diffuseFresnelCorrectionFactor(in vec3 n, in vec3 n2)
+{
+    //assert(n*n==n2);
+    bvec3 TIR = lessThan(n,vec3(1.0));
+    vec3 invdenum = mix(vec3(1.0), vec3(1.0)/(n2*n2*(vec3(554.33) - 380.7*n)), TIR);
+    vec3 num = n*mix(vec3(0.1921156102251088),n*298.25 - 261.38*n2 + 138.43,TIR);
+    num += mix(vec3(0.8078843897748912),vec3(-1.67),TIR);
+    return num*invdenum;
+}
+
+#endif
+)";
+    }
+    static std::string getLambert(const std::string&)
+    {
+        return
+R"(#ifndef _IRR_BSDF_BRDF_DIFFUSE_LAMBERT_INCLUDED_
+#define _IRR_BSDF_BRDF_DIFFUSE_LAMBERT_INCLUDED_
+
+float irr_glsl_lambert()
+{
+    return 1.0/3.14159265359;
```

Comment:
use and provide a built-in/constant for `PI` and `RECIPROCAL_PI` (`irr_glsl_` prefix required too)

Reply:
built it more like
https://github.com/buildaworldnet/IrrlichtBAW/pull/410#discussion_r368489555

I want the built-in BRDF/BSDF functions to provide the `irr_glsl_*model_name*_cos_eval` function and 2 others (build more)

Reply:
won't it be problematic when every function has NdotL inside? For example in case of cook-torrance-like thing, when both ndf and geometric term includes NdotL, then i'm getting specular factored by NdotL^2. Sorry, i dont get your idea about cos factor inside, could u explain more?

Reply:
The lighting integral (for a point) is
`Integrate_over_Hemisphere{IncomingLightDistribution(lightDirection)*BRDF(observer,lightDirection,parameters)*cos(lightDirection),over all light directions}`

The idea is that we split the integrand into two parts, one of which is the BRDF and foreshortening term (NdotL), the other is the light distribution.

We want to split that integral in 2 not 3 so `cos(theta)` has to go together with something, either the Light Distribution or the BRDF.

This is a good heuristic for Multiple Importance Sampling, as we'll be sometimes sampling (picking directions) the points on area lights and sometimes the BRDFs (directions with highest values).

Problem is that lights are well defined shapes which don't have nice contours on a hemisphere, and its hard to compute their areas (solid angles) as a function of (u,v) coordinates so its insanely hard to invert such a function to get uniform sampling. So just Importance Sampling an integrand close to 1 is near impossible for most shapes because of the contours of the integration domain, let alone when an integrand is cos(theta).

So far Triangles without the foreshortening (projection) are the only shape you can Importance Sample on a sphere under arbitrary transformations.

Furthermore Squares, Elliptical Disks are only sampleable without a transform that includes shear.

Finally Cylinders and Spheres are only sampleable without a transform that includes shear or non-uniform scale.

Only one paper mentions sampling "projected" spherical caps (with the `cos` term).


So overall it makes more sense to stick the `cos` together with the BRDF, especially since for Cook-Torrance there is already a `4*NdotL*NdotV` in the denominator, so it will cancel out.

Reply:
is it clear?

Reply:
kinda, i guess, check out what i have now (blinn_phong, lambertian, oren_nayar cos_eval functions). I dont really know what is this incoming light distribution part. Also i guess u'd like more `cos_eval` functions as builtins, i dont know which ones

Reply:
> I dont really know what is this incoming light distribution part. 

ITs the shape all the lights make or an environment map (for IBL) basically 

BRDF + cos functions I want:
- Lambertian
- Orren Nayar
- Smooth Specular (basically delta distribution given by `reflect` weighted by complex IoR Fresnel)
- Normalized Phong with Fresnel
- GGX with Correlated Smith Shadowing (same stuff as BRDF eXplorer)
- Beckmann with the best geometry term you can find
- Ashkhim Shirley anisortopic with the best geometry term you can find

BSDF + cos functions (not in this PR):
- Smooth Dielectric (Basically two delta distributions with centers given by `reflect` and `refract`, mixed together by the fresnel)
- Thin Dielectric (No refraction, but need to account for multiple fresnel and color shift)
- Rough Dielectric with GGX
- Rough Dielectric with Beckmann
- Rough Dielectric with Ashkhin Shirley

P.S. for Ahskmin Shirley anisotropic use roughnesses (invert them like in BRDF explorer), and for Phong invert the roughness into the specular exponent.

Reply:
>Smooth Specular (basically delta distribution given by reflect weighted by complex IoR Fresnel)

what is delta distribution?

Reply:
# Was is Das

https://en.wikipedia.org/wiki/Dirac_delta_function#As_a_distribution

Basically its 0 everywhere except at the center, at the center its undefined (you can return INF if argument==center).

As a probability distribution it means there is 0 probability of getting any other random value than the center one, so your `sample` BRDF function should always return the center coordinate with a probability weight of `max(cos(theta),0.0)` (original BRDF / probability of sampling = delta/delta = 1, but then remember its `brdf_cos`).

The most important part is the sampling property, you stick it as a product with another function inside any integral, the solution is the other function's value at the center of the dirac delta (without any gain or loss).

This means that a mirror is a delta distribution centered around the reflected view-vector, so it basically tells you to sample the light distribution in only one place (the exact reflection).

A smooth dielectric is two deltas, one at reflection vector, one at refraction vector, but they obviously have to be weighted together by the fresnel (not that it matters for the eval function).

# More In Depth About Sampling Multiple Deltas

However when sampling, you need to compute fresnel, and pick either to reflect or refract based on the input random value (use only one dimension).

You can use the symmetry property of the BSDF and notice that it has to have the same values when you switch the observer and the light positions around, so it is possible to "find" the centers of the deltas given a viewer position.

For a dielectric, there is the reflection where`NdotH` has to be 0, so H=N, and therefore L=-V

The refraction you can also calculate in reverse (I'd do your own function instead of calling `refract`).
https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/refract.xhtml

Due to symmetry, you can know the fresnel on a smooth dielectric by just knowing the viewer position (fresnel of `NdotV` will be the same as fresnel of `NdotL`, since V fixes reflective and refractive L vectors, and reflective fresnel fixes the transmittance).

Best choice is to pick to refract when `randomSample.x>Fresnel` then the probability of picking the refracted ray is `1.0-Fresnel` and reflected is `Fresnel` which is exactly the same as their weights in the BSDF meaning that for the probability weight of the sampling function they cancel out and you return `abs(cos(theta),0.0)` (for BSDFs the foreshortening term is abs not max).